Modules/pyexpat.c

```diff
@@ -1631,101 +1626,140 @@ static int init_handler_descrs(void)
-    if (ErrorObject == NULL) {
-        ErrorObject = PyErr_NewException("xml.parsers.expat.ExpatError",
-                                         NULL, NULL);
-        if (ErrorObject == NULL)
-            return NULL;
+    state->error = PyErr_NewException("xml.parsers.expat.ExpatError",
+                                        NULL, NULL);
+    if (state->error == NULL) {
+        return -1;
+    }
+    
+    Py_INCREF(state->error);
+    if (PyModule_AddObject(mod, "error", state->error ) < 0) {
+        Py_DECREF(state->error );
+        return -1;
+    }
+    Py_INCREF(state->error);
+    if (PyModule_AddObject(mod, "ExpatError", state->error ) < 0) {
+        Py_DECREF(state->error);
+        return -1;
+    }
+    Py_INCREF(state->xml_parse_type);
+    if (PyModule_AddObject(mod, "XMLParserType",
+                           (PyObject *) state->xml_parse_type) < 0) {
+        return -1;
     }
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "error", ErrorObject);
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "ExpatError", ErrorObject);
-    Py_INCREF(&Xmlparsetype);
-    PyModule_AddObject(m, "XMLParserType", (PyObject *) &Xmlparsetype);
 
-    PyModule_AddStringConstant(m, "EXPAT_VERSION",
-                               XML_ExpatVersion());
+    if (PyModule_AddStringConstant(mod, "EXPAT_VERSION",
+                               XML_ExpatVersion()) < 0) {
+        return -1;
+    }
     {
         XML_Expat_Version info = XML_ExpatVersionInfo();
-        PyModule_AddObject(m, "version_info",
+        if (PyModule_AddObject(mod, "version_info",
                            Py_BuildValue("(iii)", info.major,
-                                         info.minor, info.micro));
+                                         info.minor, info.micro)) < 0) {
+            return -1;
+        }
     }
     /* XXX When Expat supports some way of figuring out how it was
        compiled, this should check and set native_encoding
        appropriately.
     */
-    PyModule_AddStringConstant(m, "native_encoding", "UTF-8");
+    if (PyModule_AddStringConstant(mod, "native_encoding", "UTF-8") < 0) {
+        return -1;
+    }
 
-    d = PyModule_GetDict(m);
+    PyObject *d = PyModule_GetDict(mod);
     if (d == NULL) {
-        Py_DECREF(m);
-        return NULL;
+        return -1;
     }
-    errors_module = PyDict_GetItemWithError(d, errmod_name);
+    PyObject *errors_module = PyDict_GetItemWithError(d, errmod_name);
     if (errors_module == NULL && !PyErr_Occurred()) {
         errors_module = PyModule_New(MODULE_NAME ".errors");
         if (errors_module != NULL) {
-            _PyImport_SetModule(errmod_name, errors_module);
+            if (_PyImport_SetModule(errmod_name, errors_module) < 0) {
+                return -1;
+            }
             /* gives away the reference to errors_module */
-            PyModule_AddObject(m, "errors", errors_module);
+            if (PyModule_AddObject(mod, "errors", errors_module) < 0) {
```

Comment:
Looks like there have a refleak when `errors_module` created from  `PyModule_New(MODULE_NAME ".errors")` and calling `PyModule_AddObject` failed.

Reply:
I'm still not totally happy with it, but I fixed these leaks